src/arch/amdgpu/common/dtype/mxfp.hh

```diff
@@ -0,0 +1,323 @@
+        // of 32 by the type size (sign + exponent + mantissa bits).
+        data = raw;
+        data <<= (32 - int(FMT::sbits) - int(FMT::ebits) - int(FMT::mbits));
+    }
+
+    mxfp(const mxfp& f)
+    {
+        FMT conv_out;
+        conv_out = convertMXFP<FMT, decltype(f.getFmt())>(f.getFmt());
+        data = conv_out.storage;
+    }
+
+    mxfp &operator=(const float& f)
+    {
+       data = float_to_mxfp(f);
+       return *this;
+    }
+
+    mxfp &operator=(const mxfp& f)
+    {
+        FMT conv_out;
+        conv_out = convertMXFP<FMT, decltype(f.getFmt())>(f.getFmt());
+        data = conv_out.storage;
+        return *this;
+    }
+
+    operator float() const
+    {
+        binary32 out;
+        FMT in;
+        in.storage = data;
+        out = convertMXFP<binary32, FMT>(in, mode);
+
+        return out.fp32;
+    }
+
+    constexpr static int size()
+    {
+        return int(FMT::mbits) + int(FMT::ebits) + int(FMT::sbits);
+    }
+
+    // Intentionally use storage > size() so that a storage type is not needed
+    // as a template parameter.
+    uint32_t data = 0;
+
+    FMT getFmt() const
+    {
+        FMT out;
+        out.storage = data;
+        return out;
+    }
+
+    void setFmt(FMT in)
+    {
+        data = in.storage;
+    }
+
+    void scale(const float& f)
+    {
+        binary32 bfp;
+        bfp.fp32 = f;
+        int scale_val = bfp.exp - bfp.bias;
+
+        // Scale value of 0xFF is NaN. Scaling by NaN returns NaN.
+        // In this implementation, types without NaN define it as zero.
+        if (scale_val == 0xFF) {
+            data = FMT::nan;
+            return;
+        }
+
+        FMT in = getFmt();
+        int exp = in.exp;
+
+        if (exp + scale_val > max_exp<FMT>()) {
+            in.exp = max_exp<FMT>();
+        } else if (exp + scale_val < min_exp<FMT>()) {
+            in.exp = min_exp<FMT>();
+        } else {
+            in.exp = exp + scale_val;
+        }
+
+        data = in.storage;
+    }
+
+  private:
+    mxfpRoundingMode mode = roundTiesToEven;
+
+    uint32_t
+    float_to_mxfp(float f)
+    {
+        if (std::isinf(f)) {
+            assert(std::numeric_limits<FMT>::has_infinity);
+            return FMT::inf;
+        }
+
+        if (std::isnan(f)) {
+            assert(std::numeric_limits<FMT>::has_quiet_NaN);
+            return FMT::nan;
+        }
+
+        return float_to_mxfp_nocheck(f);
+    }
+
+    uint32_t
+    float_to_mxfp_nocheck(float f)
+    {
+        binary32 in;
+        in.fp32 = f;
+
+        FMT out;
+        out.storage = 0;
+
+        out = convertMXFP<FMT, binary32>(in, mode);
+
+        return out.storage;
+    }
+};
+
+// Unary operators
+template<typename T>
+inline T operator+(T a)
+{
+    return a;
+}
+
+template<typename T>
+inline T operator-(T a)
+{
+    // Flip sign bit
+    a.data ^= 0x80000000;
+    return a;
+}
+
+template<typename T>
+inline T operator++(T a)
+{
+    a = a + T(1.0f);
```

Comment:
What this is really doing is converting a float value (1.0f here) to the MXFP type.  It will then call `T operator+(T, T)` with the MXFP type. That operator will convert to float, do the addition, and call `T operator= ` which implicitly converts back to MXFP type and return for `a` here.  This avoids needing to define 1.0 for each MXFP type and arithmetic is not performed on MXFP types directly.  Everything is converted to float first